### PR TITLE
plugins/lsp/pylsp: `propagatedBuildInputs` -> `dependencies`

### PR DESCRIPTION
**Pull request by PerchunPak**
See https://github.com/NixOS/nixpkgs/pull/327630#discussion_r1679253679

The only (patched) pylsp plugin that doesn't use `dependencies` yet is `pylsp-rope`, though it will change soon with https://github.com/NixOS/nixpkgs/pull/328410 ([pr-tracker](https://nixpk.gs/pr-tracker.html?pr=328410)).

### CODE REVIEW

**Review by MattSturgeon — APPROVED** [review]
LGTM, thanks!

Does this fix anything or is it just updating to the new code style?

@GaetanLepage if you have time, can you also review?

**Comment by MattSturgeon** [comment]
The build [fails](https://buildbot.nix-community.org/#/builders/146/builds/230) with:

> error: attribute 'dependencies' missing

**Comment by traxys** [comment]
I think @GaetanLepage said he'd deal with this himself using the update action

**Comment by MattSturgeon** [comment]
This feels like another thing that'll break anyone who's nixpkgs lock is out of sync with outs...

**Comment by PerchunPak** [comment]
Yeah, I actually created this PR because of the error described in #1894. In nixpkgs' master every plugin uses `dependencies` already, so feel free to merge this when https://github.com/NixOS/nixpkgs/pull/328410 gets to unstable (this will fix the current evaluation error)

**Comment by MattSturgeon** [comment]
I'm wondering if we should have a more graceful approach than handles either the old/new attr depending on which is actually present, at least for a short transition period?

**Comment by PerchunPak** [comment]
Oh, yeah. I also just noticed that this rename probably won't be backported to 24.05

**Comment by MattSturgeon** [comment]
> Oh, yeah. I also just noticed that this rename probably won't be backported to 24.05

That's not an issue, since our `main` branch isn't intended to work on the stable channel.

File: plugins/lsp/language-servers/pylsp.nix
Comment (MattSturgeon) [inline plugins/lsp/language-servers/pylsp.nix:550]: This won't work, because `mkIf` is "smart" and evaluates even if the condition is false.

Since (unlike most conditional config) we _don't_ want to evaluate in this case, we should use an `if then else` wrapper such as `lib.optionalAttrs` instead.

Also, I don't think `mkMerge` and `mkIf` can be used in an override function, since they are only intended for use with module config definitions.

Something like this should work:

```nix
(
  old:
  # Note: Depending on the exact nixpkgs rev, we either need to filter `dependencies` or `propagatedBuildInputs`
  # TODO: Only filter `dependencies`
  lib.optionalAttrs (old ? propagatedBuildInputs) (filterDependencies old.propagatedBuildInputs)
  // lib.optionalAttrs (old ? dependencies) (filterDependencies old.dependencies)
)
```

Comment (MattSturgeon) [reply plugins/lsp/language-servers/pylsp.nix:550]: Or using `genAttrs`:

```nix
(
  old:
    lib.genAttrs
    (lib.filter (n: old ? ${n}) [
      "propagatedBuildInputs"
      "dependencies"
    ])
    (n: filterDependencies old.${n})
)
```

Comment (PerchunPak) [reply plugins/lsp/language-servers/pylsp.nix]: this looks so much better!

File: plugins/lsp/language-servers/pylsp.nix
Comment (MattSturgeon) [inline plugins/lsp/language-servers/pylsp.nix:554]: Since this is done gracefully I don't see why we can't merge now?

**Review by MattSturgeon — APPROVED** [review]
LGTM, but I'll let @GaetanLepage have the final say since he's dealing with the lockfile update and the relevant upstream changes?

File: plugins/lsp/language-servers/pylsp.nix
Comment (PerchunPak) [reply plugins/lsp/language-servers/pylsp.nix:554]: If you are talking about merging the PR, tests were failing. I just merged `main` to here and tests are green again

Comment (MattSturgeon) [reply plugins/lsp/language-servers/pylsp.nix:554]: Our CI is currently disabled, so the only "test" running on GitHub is `mergify` (which doesn't actually test any code).

Or are you referring to tests run locally?

Comment (PerchunPak) [reply plugins/lsp/language-servers/pylsp.nix:554]: I referred to [this test failure](https://buildbot.nix-community.org/#/builders/43/builds/287) in 6066ff2874c3b12a2418d91ca508e18340585276, and local tests fail for me even on main with some huge and totally unrelated error.

Comment (MattSturgeon) [reply plugins/lsp/language-servers/pylsp.nix:554]: > I referred to [this test failure](https://buildbot.nix-community.org/#/builders/43/builds/287) in [6066ff2](https://github.com/nix-community/nixvim/commit/6066ff2874c3b12a2418d91ca508e18340585276),

The linked failure is #1878 (one of the reasons CI is currently disabled).

> local tests fail for me even on main with some huge and totally unrelated error.

Strange. I've just run the test suite locally on another PR and everything passed just fine...

Are you using the `tests` devshell command or just `nix flake check`? Are you trying to use `--all-systems`?

Comment (MattSturgeon) [reply plugins/lsp/language-servers/pylsp.nix:554]: Just ran `tests` on this branch, and everything passes for me.

Comment (PerchunPak) [reply plugins/lsp/language-servers/pylsp.nix:554]: I used `nix flake check --all-systems -v` and got these logs:

[logs.txt.gz](https://github.com/user-attachments/files/16338335/logs.txt.gz)
(too big for pastebin, so sending as a file)

Anyway, this doesn't really matter. If all tests are passing, why not merge this PR?

Comment (MattSturgeon) [reply plugins/lsp/language-servers/pylsp.nix:554]: > why not merge this PR?

As mentioned above, I'm delegating to @GaetanLepage here since he's dealing with these changes in nixpkgs.

**Comment by GaetanLepage** [comment]
Thank you very much @PerchunPak for spinning up this fix so quickly.
I should have merged it earlier to fix it instead of waiting for the upstream fix to land on `nixos-unstable`.
Now that `pylsp-rope` also makes use of `dependencies`, I have directly used it in https://github.com/nix-community/nixvim/pull/1900/commits/9f9e20205a5631474120f2a2cd10d4ff32a6abed and the problem should now be gone.

Please, feel free to report any eventual weird behavior.

**Pull request closed without merging**
